allensdk/brain_observatory/ecephys/lfp_subsampling/subsampling.py

```diff
@@ -0,0 +1,245 @@
+# opportunities.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+import logging
+
+import numpy as np
+from scipy.signal import decimate, butter, filtfilt
+
+
+logger = logging.getLogger(__name__)
+
+
+def select_channels(total_channels,
+                    surface_channel,
+                    surface_padding,
+                    start_channel_offset,
+                    channel_stride,
+                    channel_order,
+                    noisy_channels=np.array([]),
+                    remove_noisy_channels=False,
+                    reference_channels=np.array([]),
+                    remove_references=False):
+    """
+    Selects a subset of channels for spatial downsampling
+
+    Parameters:
+    ----------
+
+    total_channels : int
+        Number of channels in the original data file
+    surface_channel : int
+        Index of channel at brain surface
+    surface_padding : int
+        Number of channels above surface to save
+    start_channel_offset : int
+        First channel to save
+    channel_stride : int
+        Number of channels to skip in output
+    channel_order : np.ndarray
+        Actual order of LFP channels (needed to account for the bug in NPX extraction)
+    noisy_channels : numpy.ndarray
+        Array indicating noisy channels
+    remove_noisy_channels : bool
+        Flag to remove noisy channels
+    reference_channels : numpy.ndarray
+        Array indicating refence channels
+    remove_references : bool
+        Flag to remove reference channels
+
+    Returns:
+    --------
+    selected_channels : numpy.ndarray
+        Indices of channels to select (relative to non-remapped data)
+    actual_channel_numbers : numpy.ndarray
+        Actual probe channels in subsampled data
+
+    """
+    assert surface_channel <= total_channels
+
+    max_channel = np.min([total_channels, surface_channel + surface_padding])
+
+    selected_channels = channel_order[start_channel_offset:max_channel:channel_stride]
+
+    actual_channel_numbers = np.arange(total_channels)
+    actual_channel_numbers = actual_channel_numbers[start_channel_offset:max_channel:channel_stride]
+
+    if remove_references:
```

Comment:
seems like this function could shrink a bit if you build the mask once and then applied it.